[web-animations] first frame of "transform" animation is visible when !important style overrides the animated value

graouts opened the pull request:
#### 6dd6ecba860d6cb63c04059fb0e2b817c9ced21e
<pre>
[web-animations] first frame of &quot;transform&quot; animation is visible when !important style overrides the animated value
<a href="https://bugs.webkit.org/show_bug.cgi?id=254665">https://bugs.webkit.org/show_bug.cgi?id=254665</a>

Reviewed by Antti Koivisto.

When fixing bug 252481, we introduced a mechanism to detect properties overriding animated properties through
the cascade, primarily !important values. We achieved this by comparing the RenderStyle after animations were
applied and after the cascade was fully applied, determining that any value that changed meant it was overridden.
But this technique failed to detect the case where the animated value matched the value after the cascade was
fully applied, as would be the case if the 0% keyframe was the same as the !important style. This would happen
for instance if there is no 0% keyframe explicitly provided.

We now use a more robust approach by populating a HashSet&lt;AnimatedProperty&gt; while the cascade is built under
Style::TreeResolver::applyCascadeAfterAnimation() and returning that list of overridden properties. We then pass
that list to the KeyframeEffectStack in didApplyCascade() which will filter it for accelerated properties only and
notify relevant keyframe effects.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-5-expected.txt: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-5.html: Added.
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::cascadeDidOverrideProperties):
(WebCore::KeyframeEffectStack::didApplyCascade): Deleted.
* Source/WebCore/animation/KeyframeEffectStack.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::overriddenAnimatedProperties const):
* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleBuilder.h:
(WebCore::Style::Builder::overriddenAnimatedProperties const):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: <a href="https://commits.webkit.org/262327@main">https://commits.webkit.org/262327@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bd2627bfd6d07ce9fa9783832ca7f43dbc3fab08

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/1265 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/1299 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/1339 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/2072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/1134 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [  ~~🧪 bindings~~](https://ews-build.webkit.org/#/builders/9/builds/1251 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/1352 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/1368 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/2072 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/1275 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/1195 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/1177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/1926 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/1208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/1165 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/1142 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/7/builds/1160 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/1197 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/2258 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/30/builds/1208 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/1104 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/4/builds/1145 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/1175 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/1235 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/125 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

anttijk commented on Source/WebCore/style/StyleBuilder.h:
```diff
@@ -35,7 +35,7 @@ namespace Style {
     WTF_MAKE_FAST_ALLOCATED;
 public:
-    Builder(RenderStyle&, BuilderContext&&, const MatchResult&, CascadeLevel, PropertyCascade::IncludedProperties = PropertyCascade::IncludedProperties::All, const HashSet<AnimatableProperty>* = nullptr);
+    Builder(RenderStyle&, BuilderContext&&, const MatchResult&, CascadeLevel, PropertyCascade::IncludedProperties = PropertyCascade::IncludedProperties::All, const HashSet<AnimatableProperty>* animatedProperties = nullptr, HashSet<AnimatableProperty>* overriddenAnimatedProperties = nullptr);
```
Passing overriddenAnimatedProperties in as a parameter everywhere seems bit unnecessary and fidly. Can it just be a member in the PropertyCascade? The client would access it after `styleBuilder.applyAllProperties()` via something like `styleBuilder. overriddenAnimatedProperties()`

anttijk commented on Source/WebCore/style/PropertyCascade.h:
```diff
@@ -41,7 +41,7 @@ class PropertyCascade {
     enum IncludedProperties { All, InheritedOnly, AfterAnimation, AfterTransition };
 
-    PropertyCascade(const MatchResult&, CascadeLevel, IncludedProperties, const HashSet<AnimatableProperty>* = nullptr);
+    PropertyCascade(const MatchResult&, CascadeLevel, IncludedProperties, const HashSet<AnimatableProperty>* animatedProperties = nullptr);
```
the variable name doesn't add much

graouts replied:
Right, that was an oversight after removing the additional parameter in a previous iteration of the patch.

anttijk commented on Source/WebCore/style/PropertyCascade.cpp:
```diff
@@ -371,5 +372,12 @@ void PropertyCascade::sortDeferredPropertyIDs()
     });
 }
 
+const HashSet<AnimatableProperty> PropertyCascade::overriddenAnimatedProperties() const
+{
+    if (m_animationLayer)
+        return m_animationLayer->overriddenProperties;
+    return { };
+}
```
Maybe just return a pointer, making copy of the HashSet seems unnecessary.

graouts replied:
The problem is that the `PropertyCascade` will be gone by the time we use the `HashSet<AnimatableProperty>` as the `StyleBuilder` that owns it will be gone as well.

graouts committed:
[web-animations] first frame of "transform" animation is visible when !important style overrides the animated value
https://bugs.webkit.org/show_bug.cgi?id=254665

Reviewed by Antti Koivisto.

When fixing bug 252481, we introduced a mechanism to detect properties overriding animated properties through
the cascade, primarily !important values. We achieved this by comparing the RenderStyle after animations were
applied and after the cascade was fully applied, determining that any value that changed meant it was overridden.
But this technique failed to detect the case where the animated value matched the value after the cascade was
fully applied, as would be the case if the 0% keyframe was the same as the !important style. This would happen
for instance if there is no 0% keyframe explicitly provided.

We now use a more robust approach by populating a HashSet<AnimatedProperty> while the cascade is built under
Style::TreeResolver::applyCascadeAfterAnimation() and returning that list of overridden properties. We then pass
that list to the KeyframeEffectStack in didApplyCascade() which will filter it for accelerated properties only and
notify relevant keyframe effects.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-5-expected.txt: Added.
* LayoutTests/webanimations/combining-transform-animations-with-different-acceleration-capabilities-5.html: Added.
* Source/WebCore/animation/KeyframeEffectStack.cpp:
(WebCore::KeyframeEffectStack::cascadeDidOverrideProperties):
(WebCore::KeyframeEffectStack::didApplyCascade): Deleted.
* Source/WebCore/animation/KeyframeEffectStack.h:
* Source/WebCore/style/PropertyCascade.cpp:
(WebCore::Style::PropertyCascade::addMatch):
(WebCore::Style::PropertyCascade::overriddenAnimatedProperties const):
* Source/WebCore/style/PropertyCascade.h:
* Source/WebCore/style/StyleBuilder.h:
(WebCore::Style::Builder::overriddenAnimatedProperties const):
* Source/WebCore/style/StyleTreeResolver.cpp:
(WebCore::Style::TreeResolver::createAnimatedElementUpdate):
(WebCore::Style::TreeResolver::applyCascadeAfterAnimation):
* Source/WebCore/style/StyleTreeResolver.h:

Canonical link: https://commits.webkit.org/262327@main
